File: android/app/src/main/java/com/heartlandwiwx/streamfeeder/data/WatchProgress.kt
package com.heartlandwiwx.streamfeeder.data

const val WATCHED_SECONDS = 30.0
const val SHORT_VIDEO_MAX_SECONDS = 60.0
const val SHORT_VIDEO_FRACTION = 0.5
const val SEEK_MAX_STEP_SECONDS = 2.0
const val WALL_CLOCK_SLACK = 1.25
const val SAMPLE_INTERVAL_MS = 400L
const val PROGRESS_PERSIST_MS = 12_000L

fun meetsWatchThreshold(
    playbackSeconds: Double,
    durationSeconds: Double?,
    ended: Boolean = false,
): Boolean {
    if (ended) return true
    if (!playbackSeconds.isFinite() || playbackSeconds < 0) return false
    if (playbackSeconds >= WATCHED_SECONDS) return true
    val duration = durationSeconds
    if (duration != null && duration.isFinite() && duration > 0 && duration < SHORT_VIDEO_MAX_SECONDS) {
        return playbackSeconds >= duration * SHORT_VIDEO_FRACTION
    }
    return false
}

data class PlaybackSampleState(
    val playing: Boolean = false,
    val lastTime: Double? = null,
    val lastWall: Long? = null,
    val playbackSeconds: Double = 0.0,
)

fun createPlaybackSampler(initialSeconds: Double = 0.0) =
    PlaybackSampleState(playbackSeconds = initialSeconds.coerceAtLeast(0.0))

fun setSamplerPlaying(state: PlaybackSampleState, playing: Boolean): PlaybackSampleState =
    if (!playing) state.copy(playing = false, lastTime = null, lastWall = null)
    else state.copy(playing = true)

fun samplePlayback(
    state: PlaybackSampleState,
    currentTime: Double,
    nowMs: Long,
    playbackRate: Double = 1.0,
): PlaybackSampleState {
    if (!state.playing) return state
    if (!currentTime.isFinite()) return state
    val rate = if (playbackRate.isFinite() && playbackRate > 0) playbackRate else 1.0
    val lastTime = state.lastTime
    val lastWall = state.lastWall
    if (lastTime == null || lastWall == null) {
        return state.copy(lastTime = currentTime, lastWall = nowMs)
    }
    val dPos = currentTime - lastTime
    val dWall = ((nowMs - lastWall) / 1000.0) * rate
    val add = if (dPos > 0 && dPos <= SEEK_MAX_STEP_SECONDS) {
        maxOf(0.0, minOf(dPos, dWall * WALL_CLOCK_SLACK))
    } else {
        0.0
    }
    return state.copy(
        lastTime = currentTime,
        lastWall = nowMs,
        playbackSeconds = state.playbackSeconds + add,
    )
}

fun mergeStoredPlayback(stored: Double, local: Double): Double = maxOf(0.0, stored, local)

File: src/lib/watchProgress.ts
import type { WatchedFilter } from '../types';

export type { WatchedFilter };

export const WATCHED_SECONDS = 30;
export const SHORT_VIDEO_MAX_SECONDS = 60;
export const SHORT_VIDEO_FRACTION = 0.5;
export const SEEK_MAX_STEP_SECONDS = 2;
export const WALL_CLOCK_SLACK = 1.25;
export const SAMPLE_INTERVAL_MS = 400;
export const PROGRESS_PERSIST_MS = 12_000;

export function parseWatchedFilter(value: string | null | undefined): WatchedFilter {
	if (value === 'watched' || value === 'unwatched') return value;
	return 'all';
}

export function matchesWatchedFilter(watchedAt: string | null | undefined, filter: WatchedFilter): boolean {
	if (filter === 'watched') return Boolean(watchedAt);
	if (filter === 'unwatched') return !watchedAt;
	return true;
}

export function meetsWatchThreshold(
	playbackSeconds: number,
	durationSeconds: number | null | undefined,
	ended = false,
): boolean {
	if (ended) return true;
	if (!Number.isFinite(playbackSeconds) || playbackSeconds < 0) return false;
	if (playbackSeconds >= WATCHED_SECONDS) return true;
	if (
		durationSeconds != null &&
		Number.isFinite(durationSeconds) &&
		durationSeconds > 0 &&
		durationSeconds < SHORT_VIDEO_MAX_SECONDS &&
		playbackSeconds >= durationSeconds * SHORT_VIDEO_FRACTION
	) {
		return true;
	}
	return false;
}

export type PlaybackSampleState = {
	playing: boolean;
	lastTime: number | null;
	lastWall: number | null;
	playbackSeconds: number;
};

export function createPlaybackSampler(initialSeconds = 0): PlaybackSampleState {
	return {
		playing: false,
		lastTime: null,
		lastWall: null,
		playbackSeconds: Math.max(0, initialSeconds),
	};
}

export function setSamplerPlaying(state: PlaybackSampleState, playing: boolean): PlaybackSampleState {
	if (!playing) {
		return { ...state, playing: false, lastTime: null, lastWall: null };
	}
	return { ...state, playing: true };
}

/** Count only small forward playhead steps while PLAYING. Seeks, pauses, and buffering add 0. */
export function samplePlayback(
	state: PlaybackSampleState,
	currentTime: number,
	nowMs: number,
	playbackRate = 1,
): PlaybackSampleState {
	if (!state.playing) return state;
	if (!Number.isFinite(currentTime) || !Number.isFinite(nowMs)) return state;
	const rate = Number.isFinite(playbackRate) && playbackRate > 0 ? playbackRate : 1;
	if (state.lastTime == null || state.lastWall == null) {
		return { ...state, lastTime: currentTime, lastWall: nowMs };
	}
	const dPos = currentTime - state.lastTime;
	const dWall = ((nowMs - state.lastWall) / 1000) * rate;
	let add = 0;
	if (dPos > 0 && dPos <= SEEK_MAX_STEP_SECONDS) {
		add = Math.max(0, Math.min(dPos, dWall * WALL_CLOCK_SLACK));
	}
	return {
		...state,
		lastTime: currentTime,
		lastWall: nowMs,
		playbackSeconds: state.playbackSeconds + add,
	};
}

export function mergeStoredPlayback(stored: number, local: number): number {
	const a = Number.isFinite(stored) ? stored : 0;
	const b = Number.isFinite(local) ? local : 0;
	return Math.max(0, a, b);
}
